JavaScript/chapter02/p04_partition/avc278.js

```diff
@@ -13,69 +13,98 @@ const {
 } = require("../../lib/avc278/linkedlist");
 
 /**
- * Partitions a linked list based on the input partition, with lesser values to the left and greater values to the right
+ * A destructive function, `partition`, splits a linked list based on the input `partitionValue`, with lesser values to
+ * the left and greater values to the right.
  * @param   {LinkedListNode} head input linked list to be partitioned
  * @return  {LinkedListNode}      partitioned linked list
  *
- * For this problem, we need to store two linked lists: one for values lesser than `partitionValue`, and one for values
- * greater than `partitionValue`. We also need to store pointers for the heads of these linked lists as we need to merge
- * them at the end. Iterating once through the original linked list makes it so the runtime is O(N). We connect nodes
- * together in the less and more lists until we exhaust the original list. At this point, we connect the end of less to
- * the head of more, set `more`'s next to null and return the head of less.
+ * For this problem, we need to store pointers to four linked list nodes:
+ * `lessHead` - pointing to the head of the linked list containing values less than the partition value
+ * `lessTail` - pointing to the tail of the linked list containing values less than the partition value
+ * `moreHead` - pointing to the head of the linked list containing values more than the partition value
+ * `moreTail` - pointing to the tail of the linked list containing values more than the partition value
+ * Iterating once through the original linked list makes it so the runtime is O(N), where N is the length of the input
+ * linked list. Since we only store four pointers to nodes at one time, our additional space required is O(1).
  * Runtime: O(N)
- * Space:   O(N)
+ * Space:   O(1)
  *
  */
 const partition = (head, partitionValue) => {
-  let less;
   let lessHead;
-  let more;
+  let lessTail;
   let moreHead;
+  let moreTail;
 
   while (head !== null) {
     if (head.val < partitionValue) {
-      if (!less) {
-        less = head;
+      if (!lessTail) {
+        lessTail = head;
         lessHead = head;
       } else {
-        less.next = head;
-        less = less.next;
+        lessTail.next = head;
+        lessTail = lessTail.next;
       }
     } else {
-      if (!more) {
-        more = head;
+      if (!moreTail) {
+        moreTail = head;
         moreHead = head;
       } else {
-        more.next = head;
-        more = more.next;
+        moreTail.next = head;
+        moreTail = moreTail.next;
       }
     }
     head = head.next;
   }
-  less.next = moreHead;
-  more.next = null;
-  return lessHead;
+
+  if (!lessHead && moreHead) {
+    head = moreHead;
```

Comment:
note that JavaScript is a [pass by value](https://medium.com/nodesimplified/javascript-pass-by-value-and-pass-by-reference-in-javascript-fcf10305aa9c) language (like most managed languages. there are some exceptions, like PHP and C# where you can specify [how the parameters are passed by](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/ref#passing-an-argument-by-reference)), so L59-L63 are no-ops* and can be removed without changing the program's behavior: the effect of executing those lines won't be visible outside this function.

\* in compiler lingo, this is a case of [Dead code elimination](https://en.wikipedia.org/wiki/Dead_code_elimination): writes to variables that are guaranteed to never be read again are often removed by the compiler since there are no observable side effects of doing (or not doing) that computation. There are some cases in which the compiler will keep the code, like when the code is being compiled in non-production mode, when there is a chance that an exception is raised in the code (since that _will_ produce visible side-effects), or in interpreted languages when a debugger is attached.

Reply:
Hmm that makes sense. But if that's the case, then what I came up with initially was a hack to pass the two edge cases
- `partition(arrayToLinkedList[10], 5)`
- `partition(arrayToLinkedList[1], 5)`

In the meantime, I'll think of a way to compensate in the main section of the algorithm for these two scenarios if this is the case.